src/ray/gcs/gcs_server/gcs_node_manager.cc

```diff
@@ -410,7 +419,11 @@ void GcsNodeManager::LoadInitialData(const EmptyCallback &done) {
   RAY_CHECK_OK(gcs_table_storage_->NodeTable().GetAll(get_node_callback));
 }
 
-void GcsNodeManager::StartNodeFailureDetector() { node_failure_detector_->Start(); }
+void GcsNodeManager::StartNodeFailureDetector() {
+  // NOTE: TO AVOID HEARTBEAT BEING DEALYED BY MAIN THREAD, MAKE SURE NODE ADDING IS
+  // ALWAYS HANDLED BY ITS OWN IO SERVICE
```

Comment:
```suggestion
```suggestion
  // Note: To avoid heartbeats being delayed by main thread, make sure node addition is always
  // handled by its own IO service.
```
```